abdiff/core/run_ab_transforms.py

```diff
@@ -78,72 +76,101 @@ def run_ab_transforms(
     """
     start_time = perf_counter()
 
+    # initialize environment
     if not docker_client:
         docker_client = docker.from_env()
-
     transformed_directory_a, transformed_directory_b = create_subdirectories(
         base_directory=run_directory, subdirectories=["transformed/a", "transformed/b"]
     )
     logger.info(
         "Transformed directories created: "
         f"{[transformed_directory_a, transformed_directory_b]}"
     )
-
     run_configs = [
         (image_tag_a, transformed_directory_a),
         (image_tag_b, transformed_directory_b),
     ]
 
-    containers = []
-    for input_file in input_files:
-        source, output_file = parse_transform_details_from_extract_filename(input_file)
-        for docker_image, transformed_directory in run_configs:
-            container = run_docker_container(
-                docker_image,
-                transformed_directory,
-                source,
-                input_file,
-                output_file,
-                docker_client,
-            )
-            containers.append(container)
-
-    failed_containers = wait_for_containers(containers, timeout)
-    logger.info(f"All {len(containers)} containers have exited.")
+    # run containers and collect results
+    futures = run_all_docker_containers(docker_client, input_files, run_configs)
+    containers, exceptions = collect_container_results(futures)
+    logger.info(
+        f"Successful containers: {len(containers)}, failed containers: {len(exceptions)}"
+    )
 
+    # process results
     log_file = aggregate_logs(run_directory, containers)
```

Comment:
Looking at the updates, it seems like [containers comprise of only the successful containers ](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/37/files#diff-7a49d9dbf619fca67eb5ed5ed16b6c62440a4cc6898754cd835d2b819be74ddcR96), so the logs from the failed containers will not appear in `logs.txt`. Was that an intended change? I think we were relying on the logs to get exceptions like "AWS credentials not being passed". I'm not seeing how the updated code includes that in the logs **or** alternatively, the `logger.info` messages indicating what containers to debug/look for the exception in? 🤔 From my understanding, the old code only [detected those lower-level errors by checking the exit code of the container](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/37/files#diff-7a49d9dbf619fca67eb5ed5ed16b6c62440a4cc6898754cd835d2b819be74ddcL199). Would be happy to chat if it'll be easier to run through the code together (I'd love to do so)!

Reply:
This, is a great point!  Nice catch.  I'll dig into how we could still capture logs from containers that did not exit cleanly (perhaps via a huddle 😄 !).

Reply:
@jonavellecuerdo - I've re-requested a review from you where hopefully commit https://github.com/MITLibraries/transmogrifier-ab-diff/pull/37/commits/bfcf42abac5143ab4acbc32d6d61156174799297 addresses this oversight.  Thanks again for catching!